webrender/src/device/gfx/descriptor.rs

```diff
@@ -2,409 +2,519 @@
-                    cache_clip_layout,
-                ));
-                assert!(self.get_pool_mut(shader_group).next());
-            },
-            ShaderGroup::Brush => {
-                self.brush_pool_idx += 1;
-                if self.brush_pool_idx < self.brush_pool.len() {
-                    assert!(self.get_pool_mut(shader_group).next());
-                    return;
+        if let Entry::Vacant(v) = self.descriptor_bindings.entry(bindings) {
+            let desc_set = v.insert(self.free_sets.get_mut(&DescriptorGroup::Default).pop().expect("Out of sampler descriptor set!"));
+            let descriptor_writes = (0..MUTABLE_SAMPLER_COUNT).into_iter().map(|index| {
+                let sampler = match bound_samplers[index] {
+                    TextureFilter::Linear | TextureFilter::Trilinear => sampler_linear,
+                    TextureFilter::Nearest => sampler_nearest,
+                };
+                hal::pso::DescriptorSetWrite {
+                    set: desc_set.raw(),
+                    binding: index as _,
+                    array_offset: 0,
+                    descriptors: Some(hal::pso::Descriptor::Sampler(sampler)),
                 }
-                let mul = 2_usize.pow(self.brush_pool_idx as u32).min(4096);
-                let descriptors_per_pool = self.descriptors_per_pool * mul;
-                let (brush_layout, mut brush_layout_range) = get_layout_and_range(
-                    pipeline_requirements,
-                    BRUSH_SHADER,
-                    self.descriptor_group_id,
-                    descriptors_per_pool,
-                );
-                self.brush_pool.push(DescPool::new(
-                    device,
-                    descriptors_per_pool,
-                    brush_layout_range,
-                    brush_layout,
-                ));
-                assert!(self.get_pool_mut(shader_group).next());
-            },
-        }
-    }
-
-    pub(super) fn reset(&mut self, device: &B::Device) {
-        self.debug_pool.reset();
-
-        // Free descritor pools which were not used in the previous draw
-        while self.cache_clip_pool_idx < self.cache_clip_pool.len() - 1 {
-            let pool = self.cache_clip_pool.pop()
-                .expect("No cache clip pool found");
-            pool.deinit(device);
-        }
-        while self.brush_pool_idx < self.brush_pool.len() - 1 {
-            let pool = self.brush_pool.pop()
-                .expect("No brush pool found");;
-            pool.deinit(device);
-        }
-
-        for pool in self.cache_clip_pool.iter_mut() {
-            pool.reset()
-        }
-        self.cache_clip_pool_idx = 0;
-        for pool in self.brush_pool.iter_mut() {
-            pool.reset()
-        }
-        self.brush_pool_idx = 0;
-        self.marked_brush_pools.clear();
-        self.marked_cache_clip_pools.clear();
+            });
+            unsafe { device.write_descriptor_sets(descriptor_writes) };
```

Comment:
something to keep in mind - we could probably prepare these descriptors ahead of time of recording